src/nvim/generators/c_grammar.lua

```diff
@@ -47,7 +47,8 @@ local c_proto = Ct(
   (fill * Cg((P('FUNC_API_NOEXPORT') * Cc(true)), 'noexport') ^ -1) *
   (fill * Cg((P('FUNC_API_REMOTE_ONLY') * Cc(true)), 'remote_only') ^ -1) *
   (fill * Cg((P('FUNC_API_LUA_ONLY') * Cc(true)), 'lua_only') ^ -1) *
-  (fill * Cg((P('FUNC_API_CHECK_TEXTLOCK') * Cc(true)), 'check_textlock') ^ -1) *
+  (fill * (Cg(P('FUNC_API_CHECK_TEXTLOCK_ALLOW_CMDWIN') * Cc(true), 'check_textlock_allow_cmdwin') +
+           Cg(P('FUNC_API_CHECK_TEXTLOCK') * Cc(true), 'check_textlock')) ^ -1) *
```

Comment:
Would it make sense to instead have:

```
FUNC_API_TEXTLOCK
FUNC_API_NO_CMDWIN
```

So `FUNC_API_TEXTLOCK` has the "old" behavior of allow cmdwin. But functions that opt-in to allowing cmdwin are also marked `FUNC_API_NO_CMDWIN`.

Reply:
Do you mean `FUNC_API_NO_CMDWIN` should disallow cmdwin?

If so, I'm happy to do it that way, but people'll need to remember to add the other attribute when creating new APIs that could potentially break the cmdwin.

Reply:
> I'm happy to do it that way, but people'll need to remember to add the attribute when creating new APIs that could potentially break the cmdwin (otherwise it's a bug if they forget)

Either way is fine. Is this better:

- `FUNC_API_TEXTLOCK` implies "no cmdwin"
- `FUNC_API_CMDWIN` opts-in to allowing cmdline

Reply:
Current way looks fine. seem a bit weird to have textlock and cmdwin coupled, but the approach in your PR doesn't look like tech debt, it's easy to revisit it later.

Reply:
Yeah, admittedly it does feel a bit weird. (And it hurts my head! 😅) 

The reason they're related at all is if something cannot work in the cmdwin, it strongly implies it changes curwin/curbuf (as that is the main restrictions in the cmdwin), which also is not allowed during textlock.
The opposite is not quite as true though; if something cannot work during textlock, it might be that it only edits the buffer text, which is still allowed in the cmdwin.